### DISCUSSION TITLE
Keycloak.X 0-downtime upgradable HotRod store

### DISCUSSION THREAD

**Discussion by mhajas** [keycloak/keycloak#8571] [Ideas]
The proposal is present in the following PR: https://github.com/keycloak/keycloak-community/pull/321 as it is too large for GH discussions. 

Markdown version of the document: https://github.com/keycloak/keycloak-community/blob/081b1382dd234fb7b32daa5802ded1fdaafe9812/design/keycloak.x/storage-hot-rod.md

**Comment by vmuzikar** [comment]
Are the transactions aligned with what [Quarkus offers](https://quarkus.io/guides/transaction)?

**Reply by vmuzikar** [reply]
@pedroigor @DGuhr ^^

**Comment by vmuzikar** [comment]
The proposal mentions only Infinispan. How does 0-downtime upgrades relate to other persistence layers like RDBMS?

**Reply by sventorben** [reply]
I was asking myself the same question. Zero-downtime is a serious requirement from many of the customers I work with. They do not need an additional "layer" like a remote infinispan store when persisting their data. This simply introduces the need for an additional infinispan cluster. So, we need to run the infinispan cluster with zero-downtime in a scalable manner and may even need another remote store to persistently store the data to disk and not only keep it in memory.

Don't get me wrong here, I think is proposal is very good, consistent and addresses the right questions. But it adds a lot of complexity to the overall system architecture where a simple RDBMS could be sufficient.

**Reply by mhajas** [reply]
Thank you for the question. I see why you are confused.

Currently, we are introducing implementations of MapStorage, `HotRodMapStorage` is just one of them. Each implementation of MapStorage is independent storage (we should have HotRod and JPA ready in the upstream in December, and we have ConcurrentHashMap in-memory store available already). Each implementation is able to do the 0-downtime upgrade of its own entities.

Also, we can choose what entities store to what storage. For example, you can have clients and users stored in JPA, however, realms, roles, groups, etc. stored in Infinispan. Or whatever combination you want. 

So to your questions:
@vmuzikar: there is no relation to RDBMS except JPA store will implement the same interface (a community proposal should be soon I think). When there will be the Tree store available (community proposal should come shortly as well), we should be able to integrate stores together, for example, we can have an Infinispan store caching values obtained from JPA, but still, each of them is responsible for 0-downtime (reading older/newer, querying etc.) on its own.
@sventorben: Infinispan is not necessary for having 0-downtime upgrades of storage that uses the database and vice-versa. Each store will be responsible for 0-downtime upgrades on its own.

**Reply by vmuzikar** [reply]
Thanks for the clarification. So if I understand correctly, 0-downtime upgrades can work with any "physical" storage implementation, e.g. even just with JPA without Infinispan? How specifically that would work? I'm probably just missing something. :)

**Reply by sventorben** [reply]
@vmuzikar I don't get this either. Why wouldn't it be possible with the current implementation then?

**Reply by mhajas** [reply]
@vmuzikar The part you are missing is probably the JPA 0-downtime storage proposal I mentioned. There will be described how we plan to achieve the same as we proposed for HotRod but for JPA. Here is [JPA WIP implementation PR](https://github.com/keycloak/keycloak/pull/8615), I believe the proposal should be in progress as well.

@sventorben We specified following requirements that each storage need to fulfill to provide 0-downtime upgradability: https://github.com/keycloak/keycloak-community/blob/master/design/keycloak.x/storage-persistence.md#introduction. The current JPA store does not provide those. You can find more details in proposals that were created prior this one: https://github.com/keycloak/keycloak-community/blob/master/design/keycloak.x/storage-architecture.md and https://github.com/keycloak/keycloak-community/blob/master/design/keycloak.x/storage-persistence.md

**Reply by bjolivot** [reply]
> Don't get me wrong here, I think is proposal is very good, consistent and addresses the right questions. But it adds a lot of complexity to the overall system architecture where a simple RDBMS could be sufficient.

When talking about 0-downtime, scalability and high availability, RDBMS are not simple, and problably the less simple component these days. All downtimes we had in keycloak for years were related to RDBMS.
Removing RDBMS dependency is in top position on my wishlist for KC  :)

**Comment by vmuzikar** [comment]
Configuring remote caches via a XML file concerns me a bit with regards to images. (E.g. the current requirement for editing `standalone.xml` for config changes is really troublesome.) What will be the best practices to configure this in a cloud environment?

**Reply by mhajas** [reply]
As far as I understood, in more complicated deployments we will prefer to have caches already created and configured (with provided/custom configurations) on the Infinispan server and Keycloak will just connect to it using URL/username/password. @martin-kanis Can you confirm?

**Reply by hmlnarik** [reply]
We need to provide a reasonable defaults, ideally without the need of the end user to configure anything for themselves, and also allow for custom alteration of that configuration

**Reply by vmuzikar** [reply]
Ok, makes sense. My only concern is that we need to avoid any XML files modifications just for the sake of configuring the server.

**Comment by vmuzikar** [comment]
Will there be any framework for the process of making changes in object representations? From the design proposal it seems any change will require a lot of manual work (incl. implementing a custom wrapper object). It would be also hard to track manually what fields can we remove it what version, and what fields need to wait for the next version to be removed.

**Reply by mhajas** [reply]
This is a great question. Currently, we don't plan to do any framework for this. I am not sure we would be able to think this through properly at the moment, as each migration can be different. I would say we need to do a few migrations manually and then maybe find out some pattern of what can be generalized to some framework. 

I will be looking into 0-downtime upgrades in the following sprint so I will try to think about this a little bit more, maybe I will find out we can do something to make migrations easier even now.

**Reply by hmlnarik** [reply]
@vmuzikar Overall, the changes in the entities structure should be exceptional.

> It would be also hard to track manually what fields can we remove it what version, and what fields need to wait for the next version to be removed.

All the changes will be visible in the code (delegates in this proposal). Also note that the versions are entity versions, not keycloak versions, so tracking those should be rather easy as they would map the migrator classes (no need to track which keycloak version they appeared in).

**Reply by vmuzikar** [reply]
If anything, IMHO we should create some scheme or template how to track the entity changes and mainly versioning, even if it should be just a text file. :)

**Reply by mhajas** [reply]
@vmuzikar We plan to have some guide/Javadoc that will summarize all steps necessary to add new migration.
